### ISSUE TITLE
bug: agent metrics exposed over prometheus can be inaccurate with multiple Coderd replicas

### ISSUE THREAD

**Issue by spikecurtis** [coder/coder#8120]
Problem
======
When there are multiple coderd replicas, the agent metrics we expose over prometheus can be inaccurate.  There are two specific problems which both stem from the same underlying issue.

1. Metric timeseries from any specific Coderd may contain stale data
2. Statistical summaries across multiple Coderd timeseries may include duplicate data, making the computed statistic inaccurate

Cause
=====
Agents post metric updates over HTTP, and when there are multiple Coderd replicas, different Coderds may field the HTTP requests from the same agent at different times.

When a Coderd receives a metric update over HTTP, it updates its local metrics cache that it uses to generate prometheus scrapes.  Stale metrics time out after 2 minutes, but are published with their most recent value from the _local_ cache until then.

Consider the following simplified example, with 2 Coderds: A & B, agent X, and metric M.

At time 0, agent X pushes M = 100, and Coderd A processes the request.
At time 10, it pushes M = 200 and Coderd B processes.
At time 20, it pushes M = 300 and Coderd B processes.
At time 30, it pushes M = 400 and Coderd A processes.

The correct timeseries (with time intervals of 10)  for metric M is: [100, 200, 300, 400] 

Coderd A reports the following timeseries:  [100, 100, 100, 400]
Coderd B reports the following timeseries: [-, 200, 300, 300]

An operator might want to know the sum of metric M across all agents.
In that sum, agent X's real contribution should be [100, 200, 300, 400], 
but the sum that prometheus will compute is [100, 300, 400, 700].

Possible Solutions
==============

Persistent connections
-----------------------
In v1, agents opened a persistent websocket to Coderd, and used that to push metrics.  If only the Coderd that serves the websocket publishes the metrics, then you avoid this issue.  If the agent drops the connection and reconnects, there is no guarantee that it hits the same Coderd, so any dashboards need to average over the coderd label, but the different timeseries should not overlap in time, leading to accurate data.

Scrape agents
--------------
Prometheus could be configured to directly scrape the agents

Store in postgres
-----------------
Agent metrics could be pushed to postgres and queried from there

**Comment by mafredri** [comment]
From the suggested solution, I think **persistent connections** is the most reasonable/simple approach to solving this issue. I do have some concerns about our agent->coderd communication, though, and this seems like a problem we'll keep running into unless we solve it in a more general way. Perhaps more communication should happen over a single websocket vs. creating new connections/http requests whenever we need to tell coderd something.

**Scraping the agents** seems like it could cause some unnecessary load. Let's say we have 1000 workspaces and all of the agents require communication via coderd. I'm not sure how Prometheus would perform the polling of 1000 agents, but I suppose it'd be something like doing 1000 requests against coderd and coderd performing 1000 outbound requests/forwards to the agents.

**Storing in PostgresSQL** is a nice solution for conflict resolution, but not out of a database maintenance/performance perspective. Because the table would either _grow_ a lot, or the constant deletes/updates would lead to _trashing_ which, in bad cases, prevents PostgreSQL autovacuuming from running (or requiring tuning) leading to table bloat. We could use unlogged tables to combat bloat but those won't be available in replicated databases.

**Comment by spikecurtis** [comment]
>  I'm not sure how Prometheus would perform the polling of 1000 agents, but I suppose it'd be something like doing 1000 requests against coderd and coderd performing 1000 outbound requests/forwards to the agents.

Here I was actually thinking of Prometheus _directly connecting_ to the agents, not thru Coderd.  But, tis true we could proxy them via Coderd.

Some of the workspace metrics (where "metrics" is used in a broad sense) end up in the front end dashboard, which means we need to either persist those metrics to Postgres or have Coderd query them on demand from the agent.  It would be nice if we had a single architecture and not several systems for different kinds of metrics.

**Comment by mafredri** [comment]
> Here I was actually thinking of Prometheus _directly connecting_ to the agents, not thru Coderd. But, tis true we could proxy them via Coderd.

How would that work in practice? Would Prometheus need to connect to our tailnet and try to reach the agents (either direct or via derp)? I don't see how else we would guarantee connectivity to the agents via external tools except require admins to setup the architecture such that agents can be reached (it's fair, but is harder to work with and potentially brittle).

> Some of the workspace metrics (where "metrics" is used in a broad sense) end up in the front end dashboard, which means we need to either persist those metrics to Postgres or have Coderd query them on demand from the agent. It would be nice if we had a single architecture and not several systems for different kinds of metrics.

I definitely agree with this. I'm not convinced Postgres is the right place to store this data, but since we can't easily recruit other tools (like Redis, etc.), it may be our best option nonetheless. Another way to work around busy table bloat is to e.g. run manual VACUUMs either as-required or on a schedule.

**Comment by spikecurtis** [comment]
> How would that work in practice? Would Prometheus need to connect to our tailnet and try to reach the agents (either direct or via derp)? I don't see how else we would guarantee connectivity to the agents via external tools except require admins to setup the architecture such that agents can be reached (it's fair, but is harder to work with and potentially brittle).

I was thinking, "require admins to setup the architecture such that agents can be reached."

One thing to say is that setting up Prometheus at all is a fairly advanced move that is currently undertaken by sophisticated Coder operators.

Another point here is that when you set up Prometheus to scrape Coderd, you don't just point it at the public access URL, as that will give you inconsistent stats when there are multiple replicas.  You have to arrange for Prometheus to have direct access to the individual replicas.  In _most_ cases, the workspaces are running in the same infrastructure as Coderd (whether it's K8s, or some VM cloud), and so you've already done all the work to get Prometheus scraper running somewhere that it could also directly access the workspace agents.

To give a Kubernetes example: you just include the right `monitoring` manifest in your workspace spec, and the Prometheus operator will start scraping your workspace.

I'd argue that kind of thing is _less brittle_, and easier to do than proxying through Coderd, since we'd need to give each agent a unique metrics path, and then provide some mechanism to automate Prometheus discovering when agents come and go and configuring the scrape.

**Comment by spikecurtis** [comment]
One thing to say is that once the v2 Agent API is merged, this will probably be a lot better, since agents will have a long lived connection with a single coderd. The connection could drop due to network hiccups or a coderd going down (/restarted), so you could have occasional badness around those edges.

**Comment by kylecarbs** [comment]
@spikecurtis the issue description seems to indicate that when we have long-lived agent<->coderd connections, this will no longer be an issue. We now have that with the v2 Agent API, so, closing!
